src/useLocation.js

```diff
@@ -0,0 +1,77 @@
+import { useState, useEffect } from 'react';
+
+function getCurrentLocation() {
+  return {
+    pathname: window.location.pathname,
+    search: window.location.search,
+  };
+}
+
+const listeners = [];
+
+function useLocation() {
+  const [{ pathname, search }, setLocation] = useState(getCurrentLocation());
+
+  /** All components using the 'useLocation' hook will update. */
+  function notify() {
+    listeners.forEach(listener => listener());
+  }
+
+  function handleChange() {
+    setLocation(getCurrentLocation());
+  }
+
+  useEffect(() => {
+    window.addEventListener('popstate', handleChange);
+    return () => window.removeEventListener('popstate', handleChange);
+  }, []);
+
+  useEffect(() => {
```

Comment:
just semantics, but this useEffect could be merged with the other useEffect since they both don't have any dependencies:
```
  useEffect(() => {
    listeners.push(handleChange);
    window.addEventListener('popstate', handleChange);
    return () => {
        listeners.splice(listeners.indexOf(handleChange), 1);
        window.removeEventListener('popstate', handleChange);
    }
  }, []);
```

Reply:
You're right, just commited with suggested change.